### ISSUE TITLE
Concurrent::Map #inspect test error

### ISSUE THREAD

**Issue by voxik** [ruby-concurrency/concurrent-ruby#547]
Trying to build package for Fedora ([1] but this link will expire quite soon), I got following error:

```
  1) Concurrent::Map #inspect
     Failure/Error: expect(Concurrent::Map.new.inspect).to match(regexp)
       expected "#<Concurrent::Map:0x-7db1e810 entries=0 default_proc=nil>" to match /\A#<Concurrent::Map:0x[0-9a-f]+ entries=[0-9]+ default_proc=.*>\Z/i
       Diff:
       @@ -1,2 +1,2 @@
       -/\A#<Concurrent::Map:0x[0-9a-f]+ entries=[0-9]+ default_proc=.*>\Z/i
       +"#<Concurrent::Map:0x-7db1e810 entries=0 default_proc=nil>"
     # ./spec/concurrent/map_spec.rb:831:in `block (2 levels) in <module:Concurrent>'
```

Please note this was 32b ARM builder, so is this some wrong mapping of integers in inspect method?

[1] http://koji.fedoraproject.org/koji/taskinfo?taskID=14762876

**Comment by jdantonio** [comment]
Thanks for pointing this out. I'll update the test with a better regex (or possibly remove the test--I'm not sure it's providing much value).

**Comment by voxik** [comment]
But I'd say that the test is pointing out an error in implementation of #inspect method (at minimun), since the "0x-7db1e810" does not look healthy IMO.

**Comment by voxik** [comment]
Interesting, today I met very similarly looking issue in Action View test suite:

```
  1) Failure:
ResolverCacheTest#test_inspect_shields_cache_internals [/builddir/build/BUILD/rubygem-actionview-5.0.0.1/usr/share/gems/gems/actionview-5.0.0.1/test/template/resolver_cache_test.rb:5]:
Expected /#<ActionView::Resolver:0x[0-9a-f]+ @cache=#<ActionView::Resolver::Cache:0x[0-9a-f]+ keys=0 queries=0>>/ to match "#<ActionView::Resolver:0x818727a8 @cache=#<ActionView::Resolver::Cache:0x-7e78d8a0 keys=0 queries=0>>".
1764 runs, 4042 assertions, 1 failures, 0 errors, 0 skips
```

Please note the ```ActionView::Resolver::Cache:0x-7e78d8a0```. I am wondering if there is common source of this issues.

**Comment by pitr-ch** [comment]
@voxik hi sorry for not addressing this sooner. I could not find any to_s or inspect implementation in our code. So I am inclined to update the regexp. Could you verify that  -7e78d8a0 is indeed an allowed value on ARM 32 ?

**Comment by voxik** [comment]
So this error is obviously coming from [[1]]. And it seems to be signed/unsigned mismatch. E.g. something like this:

~~~
pry(main)> ([0x7fffffff].pack("l").unpack("l").first).to_s 16
=> "7fffffff"

pry(main)> ([0x80000000].pack("l").unpack("l").first).to_s 16
=> "-80000000"
~~~

But in reality, this should be unpacked as unsigned:

~~~
pry(main)> ([0x80000000].pack("l").unpack("L").first).to_s 16
=> "80000000"
~~~

Obviously, this is not issue for default Object#inspect, since it is using C sprintf function with pointer format [[2]].


[1]: https://github.com/ruby-concurrency/concurrent-ruby/blame/master/lib/concurrent/map.rb#L218
[2]: http://ruby-doc.org/core-2.4.1/Object.html#method-i-inspect

**Comment by voxik** [comment]
This should be way to reproduce (be careful with the memory consumption!):

~~~
$ ruby << \EOR
GC.disable
class A
  DEFAULT_OBJ_ID_STR_WIDTH = 0.size == 4 ? 7 : 14

  def inspect
    id_str = (object_id << 1).to_s(16).rjust(DEFAULT_OBJ_ID_STR_WIDTH, '0') 
    "#<#{self.class.name}:0x#{id_str}>"
  end
end
10_000_000.times { p A.new.inspect }
EOR

"#<A:0x58585428>"
"#<A:0x585852d4>"
"#<A:0x585851bc>"
"#<A:0x5858507c>"
"#<A:0x58584ec4>"
"#<A:0x58584d5c>"
"#<A:0x58584c1c>"
"#<A:0x58584adc>"

... snip ...

"#<A:0x7fff4888>"
"#<A:0x7fff47c0>"
"#<A:0x7fff46f8>"
"#<A:0x7fff4630>"
"#<A:0x7fff4568>"
"#<A:0x7fff44a0>"
"#<A:0x7fff43d8>"
"#<A:0x7fff4310>"
"#<A:0x7fff4248>"
"#<A:0x7fff4180>"
"#<A:0x7fff40b8>"
"#<A:0x-7fffc034>"
"#<A:0x-7fffc110>"
"#<A:0x-7fffc1ec>"
"#<A:0x-7fffc2c8>"
"#<A:0x-7fffc3a4>"
"#<A:0x-7fffc480>"
"#<A:0x-7fffc55c>"
"#<A:0x-7fffc638>"
^C-:10:in `p': Interrupt
	from -:10:in `block in <main>'
	from -:10:in `times'
	from -:10:in `<main>'
~~~

When the address crosses certain level, it becomes negative. Trying this with:

~~~
$ ruby -v
ruby 2.3.3p222 (2016-11-21 revision 56859) [i386-linux]
~~~

**Comment by voxik** [comment]
On the other hand, object_id appears to be signed value (see the source at [[1]]), which doesn't make a whole lot sense to me


[1]: http://ruby-doc.org/core-2.4.1/Object.html#method-i-object_id

**Comment by voxik** [comment]
I reported the issue to Ruby upstream:

https://bugs.ruby-lang.org/issues/13397

**Comment by pitr-ch** [comment]
I think this is fixed in #699 or #651
